services/venice-server/src/main/java/com/linkedin/venice/listener/grpc/GrpcStatsContext.java

```diff
@@ -0,0 +1,363 @@
+package com.linkedin.venice.listener.grpc;
+
+import com.linkedin.venice.exceptions.VeniceException;
+import com.linkedin.venice.listener.request.RouterRequest;
+import com.linkedin.venice.read.RequestType;
+import com.linkedin.venice.stats.AggServerHttpRequestStats;
+import com.linkedin.venice.stats.ServerHttpRequestStats;
+import com.linkedin.venice.utils.LatencyUtils;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import it.unimi.dsi.fastutil.ints.IntList;
+
+
+public class GrpcStatsContext {
+  /**
+   * We need to be able to record server side statistics for gRPC requests. The current StatsHandler in the Netty
+   * Pipeline maintains instance variables per channel, and guarantees that each request will be handled by the same
+   * thread, thus we cannot augment StatsHandler in the same way as other handlers for gRPC requests. We create a
+   * new StatsContext for each gRPC request, so we can maintain per request stats which will be aggregated on request
+   * completion using references to the proper Metrics Repository in AggServerHttpRequestStats. This class is almost a
+   * direct copy of StatsHandler, without Netty Channel Read/Write logic.
+   */
```

Comment:
Can we call this class `ServerStatsContext` and have the `StatsHandler` hold an instance of this new class, rather than holding each individual stat as a class property? This way both paths could use the same code without duplication.